### ISSUE TITLE
Update to galsim v1.5?

### ISSUE THREAD

**Issue by rmandelb** [LSSTDESC/imSim#77]
If we want ImSim to be able to use a version of GalSim that has the latest SiliconSensor work, RandomWalk objects, or photon-shooting through PhaseScreen PSFs, we need to update ImSim to work with the newly-released v1.5 of GalSim.  It is possible that some of the other changes will be relevant for us (treatment of chromatic objects?).  I'm pasting the release notes below for reference; this is directly from the announcement that was distributed by @rmjarvis :

We are happy to announce a new version of GalSim: Version 1.5.

Highlights of the new functionality:

- The SEDs are now more disciplined about the units, which had been a bit poorly defined previously.
- Updated the WFIRST module to be consistent with the Cycle 7 specifications.
- Added some new profiles: DeltaFunction, RandomWalk, InclinedExponential, and InclinedSersic.
- Added a new sensor interface, including a SiliconSensor that models the LSST sensors, including the brighter-fatter effect and tree rings.
- Changes some of the syntax to be more pythonic.  The older syntax should still work in almost all cases, but will generate a deprecation warning telling you how your code should be changed.  See the [CHANGELOG](https://github.com/GalSim-developers/GalSim/blob/releases/1.5/CHANGELOG.md) for the full set of deprecations.
- Added ability to use FFTW from python directly on GalSim Images.
- Added complex Images, which are used in drawKImage and InterpolatedKImage.
- Added a geometric optics option for photon shooting PhaseScreen PSFs, which is much faster.
- Many improvements to the config layer processing that we needed for DES image simulations.
- Various bug fixes and many other more minor feature additions.  Read the [CHANGELOG](https://github.com/GalSim-developers/GalSim/blob/releases/1.5/CHANGELOG.md) for the full set of changes from v1.4 to v1.5.

So far, it is only available via [direct download](https://github.com/GalSim-developers/GalSim/releases), but it will soon be available in the various other normal places (fink, macports, LSST DM stack, ...)

**Comment by cwwalter** [comment]
I think we will  need to update the DM stack package.  That is how we have been fulfilling the requirements before. @timj How/Who makes this happen?  

Of course, If we add more things (like @jmeyers314 's atmospheric raytracing ) later.  We may need a version of master anyway.  @jchiang87 can you remind me: Did we use the stack 1.4 version for the DC1 run?

**Comment by jchiang87** [comment]
For DC1, we used a modified v13_0 version, specifically the one obtained with
```
/global/common/cori/contrib/lsst/lsstDM/setupStack-13_0_DC1_noise_fix.sh

**Comment by danielsf** [comment]
@cwwalter I think I usually handle updating the version of GalSim distributed with the stack.  I need to finish up some work for the CS group getting their new catalogs to work with CatSim.  Once that is done, I will update the stack-distributed GalSim (probably by the end of the day Tuesday; is that soon enough?)

**Comment by timj** [comment]
@danielsf that will be great. Note that we are having problems building current GalSim with anaconda 3.6.3 (it's confusing two python libraries and so all the build conftests fail). Maybe the new GalSim will not have the problem.

**Comment by danielsf** [comment]
Thanks, @timj.  I saw your conversation with John on Slack.  After I update GalSim, I will test it against 3.6.3

**Comment by cwwalter** [comment]
> For DC1, we used a modified v13_0 version, specifically the one obtained with

OK, so it looks like that should have given us the 1.4 release version of GalSim.  Thanks.

**Comment by rmandelb** [comment]
Chris: 

> Of course, If we add more things (like @jmeyers314 's atmospheric raytracing ) later. We may 
> need a version of master anyway.

That doesn't necessarily follow.  If it's really important to include some newer functionality, ad it's also really important to have a tagged release, I think the GalSim leads could be persuaded to make another release soon-ish.

**Comment by cwwalter** [comment]
> That doesn't necessarily follow. If it's really important to include some newer functionality, ad it's also really important to have a tagged release, I think the GalSim leads could be persuaded to make another release soon-ish.

Yes, that is a good point. 

I was thinking not v15.0 but no reason we couldn't use a 15.1 etc.

**Comment by rmjarvis** [comment]
If it's significant new functionality, it would more likely be 1.6.0, not 1.5.1.  The third digit is reserved for bug fixes or very minor feature updates.  But we don't need to wait more than a year between feature releases like we did between 1.4 and 1.5.  So if people have a need for another feature release soon, we'll do it.

**Comment by danielsf** [comment]
Naively updating to GalSim 1.5 broke one of Data Management's packages.  I am waiting on them to tell me it's fixed before pushing the update through.

**Comment by danielsf** [comment]
It's starting to sound like none of the Data Management developers will have time to debug the failure until next month (they are approaching the end of their Fall 2017 work cycle).  How quickly/urgently do we need the LSST stack to upgrade to GalSim 1.5?

**Comment by cwwalter** [comment]
Well.. I think in the worst case it is not bad to use the stack setup and then do a user tag and setup with a local 1.5 package. The setup gives you everything you need to do the build (that is how I normal work on master on my machines).   I'm not sure how that all fits into @heather999's NERSC setup though.

**Comment by jchiang87** [comment]
@danielsf What's the status of getting GalSim v1.5 into the sims distribution?   I don't think we want to be setting up v1.5 by hand when we start running imsim for any DC2 production.

**Comment by danielsf** [comment]
Robert Armstrong at Princeton was working to fix the DM package that broke when we naively updated to v1.5.  I will ping him to see what the status is.

**Comment by danielsf** [comment]
He tells me it is now fixed.  I will try to reimplement the GalSim update today.

**Comment by danielsf** [comment]
Actually.... It looks like Bob went ahead and updated GalSim to v1.5.1, so all that needs to be done is create a new lsst_sims version that includes it.  Do we need/want that now?

**Comment by jchiang87** [comment]
Yes, please.  We've been talking about getting imsim-based sims for protoDC2 going ASAP, so having v1.5.1 available would greatly enable that effort.   Thanks, Scott!

**Comment by heather999** [comment]
and I assume we'll want this at NERSC once the next sims release is available? @jchiang87 ?  along with whatever lsst_apps weekly the sims release corresponds to.

**Comment by danielsf** [comment]
I am currently trying to just rebuild version 2.4.2.sims against w.2017.50.  If this works, it will deliver the already-installed version of lsst_sims, except built against a DM weekly that includes GalSim v1.5.1.  This is based on the assumption that there is no other new lsst_sims functionality needed for the protoDC2 runs.

**Comment by jchiang87** [comment]
I think that's correct, Scott.  Down the road, the sensor model stuff currently in the treerings branch of sims_GalSimInterface will need to be included, but for protoDC2, I don't think it will be ready on any useful time scale.

**Comment by danielsf** [comment]
@jchiang87 @heather999 `eups distrib install lsst_sims -t sims` should now get you version 2.4.2.1 of sims. This is just version 2.4.2 of sims, except that it is built against w.2017.50, which includes GalSim v1.5.1

**Comment by heather999** [comment]
@jchiang87 @danielsf I have the new sims built on cori for KNL.. one question.. I assume you'd prefer a conda env that includes the DESCQA stuff too - right?  I'm trying to move away from the current conda env and let the descqa folks use that as a playground for development. Then I'll build a new env with just the bits we really need.  For right now, this env only contains what the stack requires, no GCRCatalogs, etc..  I can fix that - just checking if I need to do that before I set you loose.

**Comment by jchiang87** [comment]
I think we will need the gcr-catalogs and DESCQA support for this distribution.  Though I haven't settled on a final implementation, I may want to be able to generate instance catalogs for protoDC2 on a per-job basis with this install to feed to the imsim runs.

**Comment by cwwalter** [comment]
1.5.1 now comes with sims 2.4.2.1

@heather999 can this be closed?

**Comment by rmandelb** [comment]
@cwwalter has anyone checked whether the updates to v1.5 necessitate any changes in how imSim talks to GalSim?

**Comment by cwwalter** [comment]
I believe that was the .1 part of 1.5.1.

I confirmed that the tests run.

**Comment by cwwalter** [comment]
(the relevant change on imSim side was actually in the lsst-GalSim interface code which was why we needed 2.4.2.1)

**Comment by rmjarvis** [comment]
I believe there are some deprecations that should now be showing up, which should be fixed eventually, but they aren't harmful for now.  They will need to be fixed before the eventual upgrade to using GalSim 2.0.

They should be quite straightforward I think.  The deprecation warning almost always tells you what the new code should be, so hopefully there won't be anything tricky to fix.

**Comment by cwwalter** [comment]
> I believe there are some deprecations that should now be showing up, which should be fixed eventually, but they aren't harmful for now. They will need to be fixed before the eventual upgrade to using GalSim 2.0.

Yes. Now we get lots of things like:

```
/Users/walter/LSST/sims_2.4.2.1/stack/miniconda3-4.3.21-10a4fa6/DarwinX86/galsim/1.5.1.lsst1/lib/python/galsim/deprecated/__init__.py:53: GalSimDeprecationWarning: (GalSimDeprecationWarning(...), 'image.getYMax() has been deprecated since GalSim version 1.5.  Use image.ymax instead.')
  warnings.warn(s, GalSimDeprecationWarning)
/Users/walter/LSST/sims_2.4.2.1/stack/miniconda3-4.3.21-10a4fa6/DarwinX86/galsim/1.5.1.lsst1/lib/python/galsim/deprecated/__init__.py:53: GalSimDeprecationWarning: (GalSimDeprecationWarning(...), 'image.getYMin() has been deprecated since GalSim version 1.5.  Use image.ymin instead.')
  warnings.warn(s, GalSimDeprecationWarning)
ç/Users/walter/LSST/sims_2.4.2.1/stack/miniconda3-4.3.21-10a4fa6/DarwinX86/galsim/1.5.1.lsst1/lib/python/galsim/deprecated/__init__.py:53: GalSimDeprecationWarning: (GalSimDeprecationWarning(...), "copy has been deprecated since GalSim version 1.5.  GSObjects are immutable, so there's no need for copy.")
  warnings.warn(s, GalSimDeprecationWarning)
```

**Comment by jchiang87** [comment]
I fixed a handful of those in the parts of the sims code Mike and I were looking at during Sprint Week.  I'll make another pass on the treerings branch of sims_GalSimInterface and deal the rest.

**Comment by jchiang87** [comment]
Those warnings are now fixed in master.  Closed by #91
